pkg/apis/batch/validation/validation.go

```diff
@@ -168,6 +168,42 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path, opts apivalidatio
 		}
 	}
 
+	if spec.PodFailurePolicy != nil {
+		containerNames := sets.NewString()
+		for _, containerSpec := range spec.Template.Spec.Containers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for _, containerSpec := range spec.Template.Spec.InitContainers {
+			containerNames.Insert(containerSpec.Name)
+		}
+		for i, rule := range spec.PodFailurePolicy.Rules {
+			if rule.Action != batch.PodFailurePolicyActionIgnore && rule.Action != batch.PodFailurePolicyActionTerminate {
+				path := fldPath.Child("template").Index(i).Child("onExitCodes").Child("action")
+				allErrs = append(allErrs, field.NotSupported(path, rule.Action, []string{string(batch.PodFailurePolicyActionIgnore), string(batch.PodFailurePolicyActionTerminate)}))
+			}
+			if rule.OnExitCodes != nil {
+				if rule.OnExitCodes.Operator != batch.PodFailurePolicyOnExitCodesOpIn && rule.OnExitCodes.Operator != batch.PodFailurePolicyOnExitCodesOpNotIn {
+					path := fldPath.Child("template").Index(i).Child("onExitCodes").Child("operator")
+					allErrs = append(allErrs, field.NotSupported(path, rule.OnExitCodes.Operator, []string{string(batch.PodFailurePolicyOnExitCodesOpIn), string(batch.PodFailurePolicyOnExitCodesOpNotIn)}))
+				}
+				if rule.OnExitCodes.ContainerName != nil && !containerNames.Has(*rule.OnExitCodes.ContainerName) {
+					path := fldPath.Child("template").Index(i).Child("onExitCodes").Child("containerName")
+					allErrs = append(allErrs, field.Invalid(path, *rule.OnExitCodes.ContainerName, "must be one of the specified container names"))
+				}
+			}
+			if rule.OnPodConditions != nil {
+				if rule.OnPodConditions.Operator != batch.PodFailurePolicyOnPodConditionsOpIn {
+					path := fldPath.Child("template").Index(i).Child("onPodConditions").Child("operator")
+					allErrs = append(allErrs, field.NotSupported(path, rule.OnPodConditions.Operator, []string{string(batch.PodFailurePolicyOnPodConditionsOpIn)}))
+				}
+			}
+			if rule.OnExitCodes != nil && rule.OnPodConditions != nil {
+				path := fldPath.Child("template").Index(i)
+				allErrs = append(allErrs, field.Invalid(path, fmt.Sprintf("podFailurePolicy rule at index %v", i), "specifying both OnExitCodes and OnPodConditions is not supported"))
```

Comment:
I think we already have scenarios like this for other types.

@liggitt, do you think this is acceptable?

Reply:
Otherwise, I think you can use `field.Forbidden` instead.

Reply:
Maybe a sensible way out is to point that the violating item is the OnExitCodes requirement (which does not contain any pointers), and say that it co-exists with OnPodConditions?

Reply:
Updated to use Forbidden for now, will update if there is a better idea.

Reply:
Invalid is probably slightly more correct than Forbidden... to make the content look right, you could put `field.Invalid` errors on the onExitCodes and onPodConditions fields, and dereference those values, rather than having one error on the parent

Reply:
Unfortunately, this also has the same issue, because OnExitCodes uses a pointer to ContainerName. Thus, the message would still look lilke this:
```
pec.podFailurePolicy[0].onExitCodes: Invalid value: batch.PodFailurePolicyOnExitCodesRequirement{ContainerName:(*string)(0xc000441cc0), Operator:"In", Values:[]int32{1, 2, 3}}: Specifying both OnExitCodes and OnPodConditions is not supported
```

Overall, maybe the initial idea of giving a descriptive value of the parent rule is the best we have so far? It allows the use of Invalid error, points to the parent, yet avoids the serialization issue. Maybe we could just put the descriptive parent value in square brackets to make it clear this is not the actual value. Say: `fmt.Sprintf("[pod failure policy rule at index %d]", i)`?

Reply:
In the meanwhile I have used json marshaling to get string representing the rule without memory locations. It allows me to use the `Invalid` error.